### PR TITLE
refactoring Value interface

### PR DESCRIPTION
**Pull request by sigidagi**
Some functionality to remove C style unions and use Typescript discriminated unions/algebraic data types.

### CODE REVIEW

File: src/models/device.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class Device extends StreamModel implements IDevice {
         }
     }
 
-    private async _createValue(params: IValue): Promise<Value> {
+    public async _createValue(params: IValue): Promise<Value> {
```
Comment (AndreasBomholtz) [inline src/models/device.ts:76]: This should still be private.

File: test/device.test.ts [hunk 2]
```diff
@@ -490,11 +490,13 @@ describe('device', () => {
...
-            step: 1,
-            unit: 'unit',
-            si_conversion: 'si_conversion',
+            number: {
+                min: 0,
+                max: 1,
+                step: 1,
+                unit: 'unit',
+                si_conversion: 'si_conversion',
+            },
```
Comment (AndreasBomholtz) [inline test/device.test.ts:499]: We can't change the interface for creating values.

File: test/device.test.ts [hunk 3]
```diff
@@ -601,8 +603,10 @@ describe('device', () => {
             permission: 'wr',
             type: 'type',
             delta: 'delta',
-            max: 10,
-            encoding: 'encoding',
+            string: {
+                max: 10,
+                encoding: 'encoding',
+            },
```
Comment (AndreasBomholtz) [inline test/device.test.ts:609]: We can't change the interface for creating values.

File: test/device.test.ts [hunk 4]
```diff
@@ -692,8 +696,10 @@ describe('device', () => {
             permission: 'r',
             type: 'type',
             delta: 'delta',
-            max: 10,
-            encoding: 'encoding',
+            blob: {
+                max: 10,
+                encoding: 'encoding',
+            },
```
Comment (AndreasBomholtz) [inline test/device.test.ts:702]: We can't change the interface for creating values.

File: test/device.test.ts [hunk 5]
```diff
@@ -772,8 +778,10 @@ describe('device', () => {
             type: 'type',
             period: '0',
             delta: 'delta',
-            xsd: 'xsd',
-            namespace: 'namespace',
+            xml: {
+                xsd: 'xsd',
+                namespace: 'namespace',
+            },
```
Comment (AndreasBomholtz) [inline test/device.test.ts:784]: We can't change the interface for creating values.

File: test/device.test.ts [hunk 6]
```diff
@@ -858,11 +866,13 @@ describe('device', () => {
...
-            step: 1,
-            unit: 'unit',
-            si_conversion: 'si_conversion',
+            number: {
+                min: 0,
+                max: 1,
+                step: 1,
+                unit: 'unit',
+                si_conversion: 'si_conversion',
+            },
```
Comment (AndreasBomholtz) [inline test/device.test.ts:875]: We can't change the interface for creating values.

File: test/network.test.ts [hunk 7]
```diff
@@ -237,10 +237,12 @@ describe('network', () => {
...
-            min: 0,
-            max: 100,
-            step: 1,
-            unit: 'c',
+            number: {
+                min: 0,
+                max: 100,
+                step: 1,
+                unit: 'c',
+            },
```
Comment (AndreasBomholtz) [inline test/network.test.ts:245]: We can't change the interface for creating values.

File: test/value.test.ts [hunk 8]
```diff
@@ -807,10 +807,12 @@ describe('value', () => {
...
-            min: 0,
-            max: 1,
-            step: 1,
-            unit: 'unit',
+            number: {
+                min: 0,
+                max: 1,
+                step: 1,
+                unit: 'unit',
+            },
```
Comment (AndreasBomholtz) [inline test/value.test.ts:815]: We can't change the interface for creating values.

File: test/value.test.ts [hunk 9]
```diff
@@ -953,11 +955,13 @@ describe('value', () => {
...
-            max: 100,
-            step: 1,
-            unit: '',
             delta: '50',
+            number: {
+                min: 0,
+                max: 100,
+                step: 1,
+                unit: '',
+            },
```
Comment (AndreasBomholtz) [inline test/value.test.ts:964]: We can't change the interface for creating values.

File: test/value.test.ts [hunk 10]
```diff
@@ -1064,10 +1068,12 @@ describe('value', () => {
...
-            min: 0,
-            max: 100,
-            step: 1,
-            unit: '',
+            number: {
+                min: 0,
+                max: 100,
+                step: 1,
+                unit: '',
+            },
```
Comment (AndreasBomholtz) [inline test/value.test.ts:1076]: We can't change the interface for creating values.

**Review by AndreasBomholtz — CHANGES_REQUESTED** [review]

File: src/models/device.ts [hunk 11]
```diff
@@ -4,16 +4,17 @@ import { PermissionModel } from './model.permission';
 import { StreamModel } from './model.stream';
 import { Model } from './model';
 import { Value } from './value';
+import { ValuePermission } from '../util/interfaces';
```
Comment (AndreasBomholtz) [inline src/models/device.ts:7]: Move this include to the other include from util.

File: test/device.test.ts [hunk 12]
```diff
@@ -622,7 +625,7 @@ describe('device', () => {
                 string: {
                     max: 10,
                     encoding: 'encoding',
-                },
+                }
```
Comment (AndreasBomholtz) [inline test/device.test.ts:628]: Missing trailing comma.

**Review by AndreasBomholtz — CHANGES_REQUESTED** [review]

**Review by AndreasBomholtz — APPROVED** [review]
It looks good :)

**Comment by sigidagi** [comment]
merged PR to develop branch

**Pull request closed without merging**